File: Data Structure/20255/Solutions/Check_if_Array_is_Max_Heap.py
class Solution:
    def isMaxHeap(self, arr):
        n = len(arr)
        for i in range(n):
            left = 2 * i + 1
            right = 2 * i + 2
            
            if left < n and arr[i] < arr[left]:
                return False
            if right < n and arr[i] < arr[right]:
                return False
        return True

def solve():
    n = int(input())
    arr = list(map(int, input().split()))
    sol = Solution()
    result = sol.isMaxHeap(arr)
    if result:
        print("YES")
    else:
        print("NO")

if __name__ == '__main__':
    solve()

File: Data Structure/20255/Solutions/Check_if_Array_is_Max_Heap.js
class Solution {
    isMaxHeap(arr) {
        const n = arr.length;
        for (let i = 0; i < n; i++) {
            const left = 2 * i + 1;
            const right = 2 * i + 2;
            
            if (left < n && arr[i] < arr[left]) {
                return false;
            }
            if (right < n && arr[i] < arr[right]) {
                return false;
            }
        }
        return true;
    }
}

function solve(input) {
    const lines = input.trim().split("\n");
    const n = parseInt(lines[0]);
    const arr = lines[1].split(" ").map(Number);
    const sol = new Solution();
    const result = sol.isMaxHeap(arr);
    if (result) console.log("YES");
    else console.log("NO");
}

const fs = require('fs');
solve(fs.readFileSync(0, 'utf-8'));




